arch/posix/core/posix_core.c

```diff
@@ -0,0 +1,556 @@
+/*
+ * Copyright (c) 2017 Oticon A/S
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * Here is where things actually happen for the POSIX arch
+ *
+ * We isolate all functions here, to ensure they can be compiled as
+ * independently as possible to the remainder of Zephyr to avoid name clashes
+ * as Zephyr does provide functions with the same names as the POSIX threads
+ * functions
+ */
+/**
+ * Principle of operation:
+ *
+ * The Zephyr OS and its app run as a set of native pthreads.
+ * The Zephyr OS only sees one of this thread executing at a time.
+ * Which is running is controlled using pc_{cond|mtx}_threads and
+ * pc_currently_allowed_thread.
+ *
+ * The main part of the execution of each thread will occur in a fully
+ * synchronous and deterministic manner, and only when commanded by the Zephyr
+ * kernel.
+ * But the creation of a thread will spawn a new pthread whose start
+ * is asynchronous to the rest, until synchronized in pc_wait_until_allowed()
+ * below.
+ * Similarly aborting and canceling threads execute a tail in a quite
+ * asynchronous manner.
+ *
+ * This implementation is meant to be portable in between POSIX systems.
+ * A table (pc_threads_table) is used to abstract the native pthreads.
+ * And index in this table is used to identify threads in the IF to the kernel.
+ *
+ */
+
+#define POSIX_ARCH_DEBUG_PRINTS 0
+
+#include <pthread.h>
+#include <stdbool.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "posix_core.h"
+#include "posix_soc_if.h"
+#include "nano_internal.h"
+#include "kernel_structs.h"
+#include "ksched.h"
+
+#define PREFIX     "POSIX arch core: "
+#define ERPREFIX   PREFIX"error on "
+#define NO_MEM_ERR PREFIX"Can't allocate memory\n"
+
+#define PC_ALLOC_CHUNK_SIZE 64
+#define PC_REUSE_ABORTED_ENTRIES 0
+/* tests/kernel/threads/scheduling/schedule_api fails when setting
+ * PC_REUSE_ABORTED_ENTRIES => don't set it by now
+ */
+
+static int pc_threads_table_size;
+typedef struct {
```

Comment:
do you really need a new type here?

Reply:
Need, no..

Reply:
The approach is to avoid new typedefs completely except for very particular cases (slist is an example). I would prefer if you could remove those and use `struct` instead.

Reply:
will do